core/src/plugins/exec/test.ts

```diff
@@ -34,45 +34,68 @@ export type ExecTest = GardenSdkActionDefinitionActionType<typeof execTest>
 execTest.addHandler("run", async ({ log, action, artifactsPath, ctx }) => {
   const startedAt = new Date()
   const { command, env, artifacts } = action.getSpec()
+  const runCommandErrors: unknown[] = []
 
-  const commandResult = await execRunCommand({ command, action, ctx, log, env, opts: { reject: false } })
+  let commandResult
+  try {
+    // Execute the test command
+    commandResult = await execRunCommand({ command, action, ctx, log, env, opts: { reject: false } })
+  } catch (error) {
+    // Store error to be thrown at the end after trying to fetch artifacts
+    runCommandErrors.push(error)
+  }
 
-  const detail = {
-    moduleName: action.moduleName(),
-    testName: action.name,
-    command,
-    version: action.versionString(),
-    success: commandResult.success,
-    log: commandResult.outputLog,
-    startedAt,
-    completedAt: commandResult.completedAt,
+  try {
+    // Try to fetch artifacts
+    await copyArtifacts(log, artifacts, action.getBuildPath(), artifactsPath)
+  } catch (error) {
+    if (runCommandErrors.length > 0) {
+      // If the test command has failed and the artifact copy has failed as well, we just log a warning for
+      // the artifact copy error since we'll trow the test command error at the end
+      log.warn(`Failed to copy artifacts: ${error}`)
+    } else {
+      throw error
+    }
   }
 
-  const result = {
-    state: runResultToActionState(detail),
-    detail,
-    outputs: {
+  if (runCommandErrors.length === 0) {
```

Comment:
A minor non-blocking nit :)

This `if`-condition can be inverted, and the `else`-block can be then unwrapped. Thus, we'll have an error-check and throw immediately if any errors are there, and the errorless code branch will be outside a long `if`-block.

Reply:
@10ko you forgot to invert the condition (also at https://github.com/garden-io/garden/pull/6532/files#diff-fa98144b93cb2ad5871e8c16c14184f17a93eac35549d003bd0e239a134af471R59) 
```suggestion
  if (runCommandErrors.length) {
```